### PR TITLE
Support timestamp/time and date json decoding

### LINKED ISSUE #3834: Support timestamp/time and date types in json decoder

### PR DESCRIPTION
**Pull request by spebern**
# Which issue does this PR close?

Closes #3834.

# Rationale for this change

# What changes are included in this PR?

Support json decoding of
- Time32
- Time64
- Timestamp
- Date32
- Date64
in raw decoder.

# Are there any user-facing changes?

The previous json reader was a little bit more flexible, because it supported passing format strings:
https://github.com/apache/arrow-rs/blob/495682aa72ffe92bbd0d6d8d93e0c00b5483ff7d/arrow-json/src/reader.rs#L603

### CODE REVIEW

**Review by tustvold — APPROVED** [review]
Thank you, looks good to me

**Comment by spebern** [comment]
Thanks for your work on the raw decoder. It is really easy to add support for new types.

File: arrow-json/src/raw/mod.rs [hunk 1]
```diff
@@ -293,6 +293,16 @@ fn make_decoder(
         data_type => (primitive_decoder, data_type),
         DataType::Float32 => primitive_decoder!(Float32Type, data_type),
         DataType::Float64 => primitive_decoder!(Float64Type, data_type),
+        DataType::Timestamp(TimeUnit::Second, _) => primitive_decoder!(TimestampSecondType, data_type),
+        DataType::Timestamp(TimeUnit::Millisecond, _) => primitive_decoder!(TimestampMillisecondType, data_type),
+        DataType::Timestamp(TimeUnit::Microsecond, _) => primitive_decoder!(TimestampMicrosecondType, data_type),
+        DataType::Timestamp(TimeUnit::Nanosecond, _) => primitive_decoder!(TimestampNanosecondType, data_type),
```
Comment (tustvold) [inline arrow-json/src/raw/mod.rs:299]:
```suggestion
        DataType::Timestamp(TimeUnit::Second, None) => primitive_decoder!(TimestampSecondType, data_type),
        DataType::Timestamp(TimeUnit::Millisecond, None) => primitive_decoder!(TimestampMillisecondType, data_type),
        DataType::Timestamp(TimeUnit::Microsecond, None) => primitive_decoder!(TimestampMicrosecondType, data_type),
        DataType::Timestamp(TimeUnit::Nanosecond, None) => primitive_decoder!(TimestampNanosecondType, data_type),
```

I think this code doesn't correctly handle timestamps, in particular the returned arrays won't have the timestamp set. 

There is also an issue with the way the Parser implementation in general handles timezones - https://github.com/apache/arrow-rs/issues/1936

I would just leave off support for now

**Comment by tustvold** [comment]
FYI added timezone support in https://github.com/apache/arrow-rs/pull/3845

**Pull request merged**